### PR TITLE
Custom Builder Support

### PR DESCRIPTION
**Pull request by ColdenCullen**
Repos may now override any part of the project's configuration, as well as provide custom 'build' and 'test' commands.

See awslabs/aws-crt-nodejs#24 for an example of a project that uses this.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: codebuild/builder.py [hunk 1]
```diff
@@ -395,32 +402,62 @@ def get_value(self, key, args, kwds):
...
+                return
+
+            configs.append(new_config)
 
-        config_arches = config.get('architectures')
-        if config_arches:
-            config_arch = config_arches.get(build_spec.arch)
-            if config_arch:
-                configs.append(config_arch)
+            config_arches = new_config.get('architectures')
```
Comment (justinboswell) [inline codebuild/builder.py]: arches lol, archs :)

File: codebuild/builder.py [hunk 2]
```diff
@@ -613,8 +650,47 @@ def _build_project(project=None, build_tests=False, run_tests=False, build_downs
...
+                "-DCMAKE_INSTALL_PREFIX=" + install_dir,
+                "-DCMAKE_PREFIX_PATH=" + install_dir,
+                # Each image has a custom installed openssl build, make sure CMake knows where to find it
+                "-DLibCrypto_INCLUDE_DIR=/opt/openssl/include",
+                "-DLibCrypto_SHARED_LIBRARY=/opt/openssl/lib/libcrypto.so",
+                "-DLibCrypto_STATIC_LIBRARY=/opt/openssl/lib/libcrypto.a",
+                "-DCMAKE_BUILD_TYPE=" + build_config,
+                "-DBUILD_TESTING=" + ("ON" if build_tests else "OFF"),
+            ]
+            _run_command("cmake", config['build_args'], compiler_flags, cmake_args, project_source_dir)
```
Comment (justinboswell) [inline codebuild/builder.py]: Shouldn't these be ```'configure_args'```?

File: codebuild/builder.py [hunk 3]
```diff
@@ -698,22 +772,28 @@ def _build_project(project=None, build_tests=False, run_tests=False, build_downs
     _mkdir(install_dir)
 
     # Build the config object
-    config = produce_config(build_spec, sources=sources, source_dir=source_dir, build_dir=build_dir)
+    config = produce_config(build_spec, os.path.join(_cwd(), "builder.json"), sources=sources, source_dir=source_dir, build_dir=build_dir)
+    if not config['enabled']:
+        raise Exception("The project is disable in this configuration")
```
Comment (justinboswell) [inline codebuild/builder.py]: disabled

File: docker-images/linux-nodejs-x64/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,23 @@
+FROM 123124136734.dkr.ecr.us-east-1.amazonaws.com/aws-common-runtime/ubuntu-16.04:x64
+
+ENV NODE_VERSION="10.1.0"
+
+###############################################################################
+# NodeJS
+###############################################################################
+RUN set -ex \
```
Comment (justinboswell) [inline docker-images/linux-nodejs-x64/Dockerfile]: Should we use nvm and install 11 and 12 and test against those too? Maybe as a follow-up?

Comment (ColdenCullen) [reply docker-images/linux-nodejs-x64/Dockerfile]: That's a great idea... for later! Let's get this rolled out to the other CRTs first.

**Pull request merged**